refactor: improve sync reliability

tyler-dot-earth opened the pull request:
# related issues

- fixes #45
- may help #68
  - unclear what underlying cause of this issue is, though this PR removes code that removes items from `booksIDsMap` altogether - so i suspect it should help as that is now modified less
  - this issue is also caused by Obsidian Sync
- may help #14

# tl;dr

- substantially improve the ability to resync deleted items
- unify resync logic
- fix issues where the plugin thought there was nothing to sync
- reduces hacks/workarounds, and generally improve the codebase

# sync changes

> [!NOTE]  
> some name changes since i originally posted the notes below, see [this comment](https://github.com/readwiseio/obsidian-readwise/pull/67#discussion_r1739794857)
> - `refreshBookExport` has been renamed to `syncBookHighlights`
> - `startSync` has been folded into `requestArchive`
> - `requestArchive` has been renamed to `queueExport`
> - `downloadArchive` has been renamed to `downloadExport`

Primarily:

1. Make `refreshBookExport` async
    - There is no reason for it not to be, though that's largely because
      it is now debounced. (keep reading)
2. Remove `debounce` from refreshBookExport
    - I suspect this only ever existed because the 'delete' listener
      fires for *each* file that is deleted, which in turn was firing
      the `refreshBookExport` function a ton of times. (it isn't called
      in the 'delete' listener anymore - keep reading)
3. Defer `refreshBookExport` to sync events
    - aka don't call this when deleting files
    - ensure `refreshBookExport` called when syncing - both on
      scheduled and on-demand
4. Generally ensure *all* syncing happens via `refreshBookExport`
    - ensures that the various sync methods won't get tangled up very easily (eg writing to settings via various sync operations in tandem)

Additionally:
- async/await *most* `saveSettings` - i suspect this was causing issues where settings would get clobbered. either way, couldn't see a reason not to do this in *most* cases.

All of these combined should ensure a much more reliably sync operation.

Extensive testing instructions found at bottom of this post ⤵

# notes dump

clean plugin install happens

default `data.json` (aka `settings`):
```
{
  "token": "",
  "readwiseDir": "Readwise",
  "frequency": "0",
  "triggerOnLoad": true,
  "isSyncing": false,
  "lastSyncFailed": false,
  "lastSavedStatusID": 0,
  "currentSyncStatusID": 0,
  "refreshBooks": false,
  "booksToRefresh": [],
  "booksIDsMap": {},
  "reimportShowConfirmation": true
}
```

user clicks "connect"
- `getUserAuthToken()` → saves to `settings.token`
- web browser opens to `https://readwise.io/export/obsidian/preferences`
- obsidian plugin UI refreshes to show all options now that user is authenticated

note that sync frequency is manual/`0` at this point

user initiates initial/first sync — this can happen one of several ways:
1. click "initiate sync"
	- `refreshBookExport`
1. `Sync your data now` command
	- `refreshBookExport`
1. toggle on "resync deleted files"
	- `refreshBookExport`
1. reload workspace
	- `onload` → `refreshBookExport`
1. interval is triggered
	- `configureSchedule` → `refreshBookExport`

after the initial sync, `settings.lastSavedStatusID` and `settings.booksIDsMap` are populated

at this point, the user may also use the `Delete and reimport this document` command within a specific Readwise export — that also simply uses `refreshBookExport`

 notes about `refreshBookExport`
- **all** syncing goes through this. ==very important== because it ensures that the various sync methods won't get tangled up very easily (eg writing to settings via various sync operations in tandem)
- `refreshBookExport` will always trigger a `requestArchive` (either directly or via `startSync`), regardless of if it was provided specific books to refresh, so new highlights should always appear

notes about "resync deleted files" (aka `settings.refreshBooks`):
- if enabled OR disabled:
 	- the entry will remain in `booksIDsMap` so we have a copy of its book ID
	- book IDs will NOT be added to `booksToRefresh` as they are deleted (see `if enabled` for rationale)
- if enabled:
	- `refreshBookExport` will look for a difference in the filesystem vs what is in `booksIDsMap` — seemed like the only way to handle filesystem deletions (outside of Obsidian), unknown performance when there are many many readwise files. could be moved behind a toggle if it's particularly slow.

other notes:
- `auto` query string seems to indicate that the sync wasn't triggered by a user (eg interval sync)
- when `settings.readwiseDir` isn't there, the `requestArchive` will use the `parentPageDeleted` querystring (like `/api/obsidian/init?parentPageDeleted=true`) — there's no documentation on this, but i did test it and it does successfully sync the files in a clean install

## install for testing

you can download and build the plugin yourself, or use [BRAT](https://github.com/TfTHacker/obsidian42-brat) to install a pre-built beta:
1. disable the current Readwise plugin in your Obsidian
1. install [BRAT](https://github.com/TfTHacker/obsidian42-brat) via Obsidian community plugins and enable the BRAT plugin
1. within BRAT's plugin settings, click the "Add Beta plugin with frozen version" button
1. fill the fields in with the beta info 
  Repository: `https://github.com/tyler-dot-earth/dev-obsidian-readwise`
  Latest pre-built version: `refactor-improve-sync-reliability-beta-5`
  Should look a bit like this:
  ![Screenshot from 2024-07-06 10-48-27](https://github.com/readwiseio/obsidian-readwise/assets/595711/7c1b8784-dcef-402c-be25-36f36afa3731)
1. use the plugin like normal


## how to test thoroughly

> [!WARNING]  
> perform these in listed order

- [ ] confirm clean plugin install syncs notes as expected (this isn't automatic; you must click "initiate sync" or reload the workspace after authenticating)

- **enable `Resync deleted files`**
- delete a synced file
- use command palette to `Readwise Official: Sync your data now`
- [ ] confirm deleted file **is** re-synced

- delete a synced file
- open plugin settings and click `Initiate sync`
- [ ] confirm deleted file **is** re-synced

- open a synced file
- use command palette to `Readwise Official: Delete and reimport this document`
- [ ] confirm file is deleted and **is** re-synced

- delete synced file
- use command palette to `Reload app without saving` (or re-open Obsidian)
- [ ] confirm deleted file **is** re-synced

- **disable `Resync deleted files`**
- delete a synced file
- use command palette to `Readwise Official: Sync your data now`
- [ ] confirm file is **NOT** re-synced

- delete a synced file
- open plugin settings and click `Initiate sync`
- [ ] confirm file is **NOT** re-synced

- open a synced file
- use command palette to `Readwise Official: Delete and reimport this document`
- [ ] confirm file **is** re-synced

- delete synced file
- use command palette to `Reload app without saving` (or re-open Obsidian)
- [ ] confirm deleted file **is** re-synced

- delete synced file
- **enable `Resync deleted files`**
- [ ] confirm deleted file **is** re-synced

- and finally, for good measure:
- [ ] ensure adding new highlights continues to work as expected
- [ ] ensure syncing on interval continues to work as expected


# other changes

- a few random, small bugfixes and refactors. i'm happy to split into their own PR, but felt like more friction than it was worth for both me and reviewer(s).

TristanH commented:
> Remove debounce from refreshBookExport
I suspect this only ever existed because the 'delete' listener
fires for each file that is deleted, which in turn was firing
the refreshBookExport function a ton of times. (it isn't called
in the 'delete' listener anymore - keep reading)

I'd definitely like us to confirm this... @tadeoos do you recollect the original rationale?

tyler-dot-earth commented:
@TristanH 
> > Remove debounce from refreshBookExport
> > I suspect this only ever existed because the 'delete' listener
> > fires for each file that is deleted, which in turn was firing
> > the refreshBookExport function a ton of times. (it isn't called
> > in the 'delete' listener anymore - keep reading)
> 
> I'd definitely like us to confirm this... @tadeoos do you recollect the original rationale?

Please do! I could only guess based on the code and behavior.

Given the debounce was 800ms, I can't imagine it was to compensate for some backend behavior (eg a slow sync process that would get confused by "outdated" `refreshBookExport` request)

The relevant changes call `refreshBookExport` *significantly* less as it is simply no longer called immediately upon deletion of *every single file*, instead deferring to the normal sync schedule or manual re-sync.

So beyond spamming the request, which this PR *significantly reduces*, i suppose the question to answer is: is there any point in **keeping** the debounce?

That said, I can also say:
- this syncing behavior is working really well from my testing
- `requestSync` was never similarly debounced (though perhaps that's a problem of its own)
- rate limiting may be a more generalized solution to this problem

TristanH commented on src/main.ts:
```diff
@@ -352,34 +354,34 @@ export default class ReadwisePlugin extends Plugin {
         {
           headers: {...this.getAuthHeaders(), 'Content-Type': 'application/json'},
           method: "POST",
           body: JSON.stringify({exportTarget: 'obsidian', books: bookIds})
```
@tyler-dot-earth we received some requests to this endpoint on our backend with the `books` key undefined -- this would have returned a 500

it was at ~12:42AM UTC.

was that just you messing around with stuff, or potentially a new regression on this PR?

tyler-dot-earth replied:
@TristanH I was working on a last lil bit of optimizatio. Lemme do a bit more testing to verify, will respond back soon 👍

tyler-dot-earth replied:
@TristanH Pretty sure the 500s were from WIP code that I hadn't committed. That said, here's some updates:

- i added a *HUGE* list of test cases to the original post so it can be thoroughly tested by others
- i added a commit to straighten out sync/refresh behavior better control when books are removed from the list of book IDs to refresh
- i added a commit to restore missing files when "Resync deleted files" is enabled
  - not 100% happy with the solution (find missing books from bookIdsMap that aren't in the vault), but it was the easiest way to achieve this
  - ^ the core of my issue is that the codebase doesn't really make a distinguishment between books to refresh and books which have been deleted
  - ... so now books that have been deleted do NOT get added to the fresh list; refresh list is now just used as the queue for syncs (meaning deleted files won't end up in that queue unless user has `resync deleted files` enabled)

i can attempt to further explain the rationale as needed (or further refine the logic/code), but it's getting late at the moment.

take a look and try the test cases when you have a bit of time (: i'll run through them again after i sleep to double check everything.

tyler-dot-earth replied:
Noticing clean install sync isn't working as expected; moving this back draft while I fix. Will update soon.

tyler-dot-earth commented:
@TristanH alright, i tested the heck out of this today — behavior seems solid, and I think I fixed a few more related bugs along the way. 

added a ton of notes to the original post, and a big list of steps for testing.

enjoy!

tyler-dot-earth commented:
added an `install for testing` section to OP with instructions on testing via [BRAT](https://github.com/TfTHacker/obsidian42-brat) instead of having to manually clone & install.

tyler-dot-earth commented:
Noticing that `null` is being added to `books to refresh`.

Not necessarily a problem - this happened before, too - but will see if there's something specific causing it. Would be easy to filter them out, but would be better if it didn't happen to begin with.

![image](https://github.com/readwiseio/obsidian-readwise/assets/595711/eb0fccf1-b09c-4249-9488-6816d4b10d81)

tyler-dot-earth commented on src/main.ts:
```diff
@@ -388,95 +433,76 @@ export default class ReadwisePlugin extends Plugin {
-      if (bookId) {
+      // user has "Resync deleted files" enabled
+      if (this.settings.refreshBooks) {
+        const bookId = this.settings.booksIDsMap[file.path];
+        if (!bookId) return;
+
+        // queue the book for refresh
         await this.addBookToRefresh(bookId);
       }
-      this.refreshBookExport();
-      delete this.settings.booksIDsMap[file.path];
-      this.saveSettings();
     });
```
i'm not convinced that this `on(delete)` listener is needed at all with my other changes.

- entries are no longer removed from `booksIDsMap`
- `refreshBookExport` will now look for the delta between `booksIDsMap` and your actual system
- ... and if `refreshBooks` (re-sync deleted files) is enabled, will sync those missing files

tyler-dot-earth replied:
i deleted the `on(delete)` listener and verified that it is now unnecessary by running through the test set in OP

TristanH reviewed (commented):
awesome work on this man. hopefully @tadeoos can have a thorough review next week

in the mean time, we've been testing it internally and so far so good!

also really appreciate you sharing it on the reddit and making the testing process easy 😍

tyler-dot-earth commented:
@TristanH i appreciate the support! and very happy to hear that you guys have been testing it internally without issue.

i pushed one final change removing the `on(delete)` listener [per my previous comment](https://github.com/readwiseio/obsidian-readwise/pull/67#discussion_r1674078706) that it now serves no purpose. with that, **beta 4 is now tagged**.

i'll keep a look out for review comments :smiley:

TristanH commented:
@homostellaris in general we are pretty game to merge this

last i heard (@tyler-dot-earth correct me if i'm wrong) we didn't actually think that the changes here would solve a lot of the syncing issues linked (which are usually caused by using Obsidian Sync)

but if we think they will substantially help, definitely game to merge them

tyler-dot-earth commented:
@TristanH While this PR won't resolve the "race condotion" of Readwise sync vs Obsidian Sync which is clobbering book IDs (#68), it does substantially improve the ability to resync deleted items, unify resync logic, fix issues where the plugin thought there was nothing to sync, reduces hacks/workarounds, and generally improve the codebase.

I could have made those improvements more clear in my OP — my bad.

I've been using it this whole time with no issue aside from the (preexisting) book IDs issue that is caused by the aforementioned race condition. I have ideas about how to improve/fix that (#68), though hesitant to start on them until these fixes/restructure is approved 🙂

TristanH commented:
@tyler-dot-earth great! @tadeoos should be able to do one final code review tomorrow, and then I think we'll be good to get it merged 👍 

not your fault at all, i should have asked

tadeoos commented on src/main.ts:
```diff
@@ -345,41 +358,73 @@ export default class ReadwisePlugin extends Plugin {
+    const targetBookIds = bookIds || this.settings.booksToRefresh;
+
+    // add potentially-missing books to booksToRefresh (TODO - prob a lil inefficient? 🤷)
+    const knownFilesPaths = Object.keys(this.settings.booksIDsMap);
```
Do I understand correctly that the code below is why you may not need the `app.vault.on("delete")...` handler?

tyler-dot-earth replied:
Correct :heavy_check_mark: This means that the `booksIDsMap` stays more consistent as it always representing *all* known book IDs which makes re-syncing deleted files easier - just look at the delta of filesystem vs `booksIDsMap`.

Fuzzy on this, but also believe I think `on("delete")` had potential to behave in unexpected ways if you deleted many files at once (eg a whole folder) because you could not ensure that its operations finished before the next `on("delete")` fires -- so the operations within this would compete with each other when modifying the `data.json` & saving.

tadeoos replied:
Niice

tadeoos commented on src/main.ts:
```diff
@@ -388,95 +433,66 @@ export default class ReadwisePlugin extends Plugin {
-    this.app.vault.on("delete", async (file) => {
-      const bookId = this.settings.booksIDsMap[file.path];
-      if (bookId) {
-        await this.addBookToRefresh(bookId);
```
I think this is the only place where we use `addBookToRefresh` so if this goes, the function should also go, albeit it's not yet clear to me we can remove it fully...

tyler-dot-earth replied:
`addBookToRefresh` is used `catch` of in `downloadArchive` when it fails to write.

In theory... the book ID may not exist when the `catch` is triggered as the book ID is set partway through the `try`.

So it's possible that we could still probably be purged if we pull the name/bookID splitting up out of the `try/catch` and then save the book ID to the settings. In fact, I think it's probably a good idea to refactor this part of the code, otherwise you might end up with `undefined` in the `booksToRefresh`.

I'll investigate this after more critical parts of feedback.

tyler-dot-earth replied:
OK, so I don't think `addBookToRefresh` can be eliminated because it covers the case where the user **doesn't** have "resync deleted books" enabled **AND** an entry couldn't be saved to the filesystem.

That said, I still reworked the code a little:
- 62f2744 (logic rework)
- bd6f30c (add comment about use of `addBookToRefresh`)

tadeoos commented on src/main.ts:
```diff
@@ -345,41 +358,73 @@ export default class ReadwisePlugin extends Plugin {
+
+    if (!targetBookIds.length) {
+      console.log('Readwise Official plugin: no targetBookIds, triggering check for other updates...');
+      await this.requestArchive(null, null, auto);
```
hmm... this won't set the `isSyncing` setting that happens via `startSync` function... did u miss this or for some reason it's not a problem?

tyler-dot-earth replied:
Looks like an oversight, but I think I have a solution here: https://github.com/readwiseio/obsidian-readwise/pull/67#discussion_r1739794857

tyler-dot-earth replied:
I did the proposed solution in the previous comment :arrow_heading_up: so the `isSyncing` should now getting set here :cowboy_hat_face:

tadeoos commented on src/main.ts:
```diff
@@ -345,41 +358,73 @@ export default class ReadwisePlugin extends Plugin {
-    if (!bookIds.length || !this.settings.refreshBooks) {
+  /** Initiates a sync, or queues refresh for later if queue fails.
+   * ALL syncing goes through this function. */
+  async refreshBookExport(
```
heh... The decision to make this the only entry point is good codewise, but it makes the name of this function a lil bit outdated... "Refreshing book export" is meaningful in our backend export flow and it is very narrow: it marks some book to be for sure included in the next export. Clearly this function does this only at the end and can return earlier :) So what would be the better name... `triggerSync`? Too close to `startSync`... Maybe `initiateReadwiseSync` ?

tyler-dot-earth replied:
What do you think of this name/change? fcdfe0e

Thinking is that `syncBookHighlights` -> `startSync` ... though maybe there's even more room for improvement now that I'm writing it out.. Would be nice if the whole flow, like
`syncBookHighlights` -> `queueExport` -> `getExportStatus` -> `downloadExportArchive`

tyler-dot-earth replied:
:arrow_up: I went ahead and made this change for the sake of async comms - easy to undo/amend if you're not into it.

tadeoos replied:
Niiiiice 👍🏻

tadeoos reviewed (changes requested):
Dude... Really nice work! 

Thanks so much for the PR and thanks a ton for your patience as I was focused on some high prio stuff (completely unrelated to Obsidian) and finally got time to review this just today. Also: thanks for making my old typescript code suck less :)

I want to move responsibly here as I haven't touched this in a couple of years so I don't remember all of the nitty-gritty details right away — I left some comments to make sure I understand your intentions and pointed out some places where you may have missed some UX corner cases...

But all in all, I think we're close to shipping this, hopefully early next week!

tadeoos commented:
One more thing:
(from your notes):
> notes about "resync deleted files" (aka settings.refreshBooks):
> if disabled:
> the entry will remain in booksIDsMap so we have a copy of its book ID
> book IDs will NOT be added to booksToRefresh
> if enabled:
> book IDs will be added to booksToRefresh as they are deleted
> the entry will remain in booksIDsMap
> refreshBookExport will look for a difference in the filesystem vs what is in booksIDsMap — seemed like the only way to > handle filesystem deletions (outside of Obsidian), unknown performance when there are many many readwise files. could be moved behind a toggle if it's particularly slow.

I'm not sure I see this logic fully implemented in your current version... We now only add books to `booksToRefresh` in one place (`downloadArchive`) and there is no setting check there.

tyler-dot-earth commented:
> I'm not sure I see this logic fully implemented in your current version... We now only add books to `booksToRefresh` in one place (`downloadArchive`) and there is no setting check there.

Ah, yes - I think I wrote that before I made the changes to the deletion logic which obsolesced the need for storing the deleted IDs in the `booksToRefresh`.

tyler-dot-earth commented:
> I left some comments to make sure I understand your intentions and pointed out some places where you may have missed some UX corner cases...

Thank you for the review @tadeoos :)

I have given an initial look-through, good feedback - I should hopefully have some time this weekend to really dig in and make updates.

tyler-dot-earth commented on src/main.ts:
```diff
@@ -388,95 +433,66 @@ export default class ReadwisePlugin extends Plugin {
-    );
-
-    this.refreshBookExport(this.settings.booksToRefresh);
-    this.app.vault.on("delete", async (file) => {
```
GOTO https://github.com/readwiseio/obsidian-readwise/pull/67#discussion_r1739756049

tyler-dot-earth commented on src/main.ts:
```diff
@@ -388,95 +433,66 @@ export default class ReadwisePlugin extends Plugin {
       );
     }
-    await this.loadSettings();
-    this.refreshBookExport = debounce(
```
Yup. `loadSettings` is moved up to the top of `onLoad` and `debounce` no longer needed because (i think) it was mostly being fired by some excessive calls that were happening elsewhere (the `on(delete)`, maybe?) + more thorough `async`/`await`.

tyler-dot-earth commented on src/main.ts:
```diff
@@ -388,95 +433,66 @@ export default class ReadwisePlugin extends Plugin {
-      });
+      this.notice("Deleting and reimporting file...", true);
+      await vault.delete(vault.getAbstractFileByPath(fileName));
+      const bookId = this.settings.booksIDsMap[fileName];
```
I think it's highly unlikely as there is a `checkCallback` when the reimport command is added via `addCommand` (which means they can't even use that command unless they're in a readwise file), though no real harm in adding a check if you'd prefer.

tyler-dot-earth commented on src/main.ts:
```diff
@@ -388,95 +433,66 @@ export default class ReadwisePlugin extends Plugin {
-        }
-      });
+      this.notice("Deleting and reimporting file...", true);
+      await vault.delete(vault.getAbstractFileByPath(fileName));
```
Hmm. It's a good observation - but digging in a little, here's some good info to consider:

All `/api/refresh_book_export` does is *queues* the export. So, the old code wasn't actually waiting for the file to actually be reimported, it was queuing and *assuming* the archive gets downloaded/extracted/etc.

The code in this branch makes just 1 more assumption: that queuing will be successful.

I think it's an OK assumption with that framing - also remembering: deleting a file just puts it in the trash, the user can always kick the readwise sync off again, or the user can utilize their sync solution to restore it.

If you disagree, I'm happy to figure this out before merging. Alternatively, we could open an issue to track that files may be deleted before successful reimport and tackle it separately (happy to take responsibility on that).

tadeoos replied:
That's a good point, let's go with optimism :)

tyler-dot-earth commented on src/main.ts:
```diff
@@ -388,95 +433,66 @@ export default class ReadwisePlugin extends Plugin {
+      this.notice("Deleting and reimporting file...", true);
+      await vault.delete(vault.getAbstractFileByPath(fileName));
+      const bookId = this.settings.booksIDsMap[fileName];
+      await this.refreshBookExport([bookId]);
```
How about this mod? Only get deleted books if no book IDs are explicitly passed in.

2baf6bc
```
    const shouldGetMissingBooks = this.settings.refreshBooks && !bookIds?.length;
    if (shouldGetMissingBooks) {
```

tyler-dot-earth commented:
@tadeoos Did a round of fixes and responses to your review. Also bumped the BRAT version to `refactor-improve-sync-reliability-beta-5` for some dogfooding.

tadeoos reviewed (approved):
Thanks a ton for a quick iteration here, great changes!
